### PR TITLE
[genesis] migrate validator and operator initialization to Move

### PR DESCRIPTION
**Pull request by sblackshear**
Genesis contains rather intricate logic for initializing the genesis validator and operator set. This top commit of the PR migrates this logic into Move to facilitate testing and verification (along the lines of what @DavidLDill has been requesting). 

This also opens the door for simplifying the caller API for specifying validator assignments (second commit). Currently, the caller of `vm_genesis_encode_genesis_transaction` must pass two vectors of type `(Option<Ed25519PublicKey>, Name, ScriptFunction)` and `(Ed25519PublicKey, Name, ScriptFunction)`. The structure and relationship of these vectors is fairly complex. This PR simplifies the API so the caller only needs to pass a single vector of `Validator` objects that contain the necessary data about a validator and its operator.
    
This simplification lets us delete a lot of code in both the config builder and vm_genesis, and kills a dependency on `diem_transaction_builder` in the config builder as well.

### CODE REVIEW

File: language/diem-framework/modules/Genesis.move [hunk 1]
```diff
@@ -89,6 +93,86 @@ module Genesis {
         DiemTimestamp::set_time_has_started(dr_account);
     }
 
+    fun create_initialize_owners_operators(
```
Comment (DavidLDill) [inline language/diem-framework/modules/Genesis.move:96]: I'd love a comment explaining this.  I had to get well into the function to understand that "owners" is a vector of validator owners.   I'll have to finish reading the function then look for where it's called to understand where the args come from.

**Review by DavidLDill — APPROVED** [review]

File: language/diem-framework/modules/Genesis.move (hunk 1 above)
Comment (sblackshear) [reply language/diem-framework/modules/Genesis.move:96]: Good idea, I had to do a lot of investigating to understand the old Rust code, will try to encode that knowledge in comments.

**Review by runtian-zhou — DISMISSED** [review]
Refactor looks clean and good to me. Genesis.move module will be filtered out in production release so this will not hurt the actual Diem deployment whatsoever. There are three locations where genesis is involved in CI pipeline:
1. Functional tests: we will test the validator assignment operation in move-lang's diem associated functional tests. This will prove that the new validator assignment workflow will function properly.
2. e2e-testsuite: there's only one test case in e2e-testsuite that uses fresh genesis, which is single p2p. So it shouldn't be affected at all.
3. cluster test: The backward compatibility testsuite will generate a genesis and start a network using the new genesis tooling. This will prove that we will be able to spawn a new network with the new flow.

**Comment by sblackshear** [comment]
Thanks for the context!

> cluster test: The backward compatibility testsuite will generate a genesis and start a network using the new genesis tooling. This will prove that we will be able to spawn a new network with the new flow.

A question: how strict are these tests? Are they just checking that we are able to initialize a node without crashing using the new genesis, or are they also testing that replaying produces the "same" results as the old genesis (for some definition of "same": same root hash, same write set for each tx, ...).

**Comment by runtian-zhou** [comment]
> Thanks for the context!
> 
> > cluster test: The backward compatibility testsuite will generate a genesis and start a network using the new genesis tooling. This will prove that we will be able to spawn a new network with the new flow.
> 
> A question: how strict are these tests? Are they just checking that we are able to initialize a node without crashing using the new genesis, or are they also testing that replaying produces the "same" results as the old genesis (for some definition of "same": same root hash, same write set for each tx, ...).

It's the former one. It's simply checking the node initialization is fine.

**Pull request closed without merging**